File: 2021/15/202115.py
from timeit import timeit
from typing import List, Tuple


def parse(s: str) -> List[List[int]]:
    return [[int(c) for c in line]
            for line in s.splitlines()]


td = parse("""
1163751742
1381373672
2136511328
3694931569
7463417111
1319128137
1359912421
3125421639
1293138521
2311944581
""".strip())
data = parse(open("input.txt").read())

INF = 999999999

RT, DN, LT, UP = 0, 1, 2, 3


def move(r: int, c: int, dr: int) -> Tuple[int, int]:
    if dr == RT:
        return r, c+1
    elif dr == DN:
        return r+1, c
    elif dr == LT:
        return r, c-1
    else:
        return r-1, c


def search(d: List[List[int]]) -> int:
    nr, nc = len(d), len(d[0])
    mn = [([INF] * nc) for _ in range(nr)]
    sr, sc = 0, 0
    st = {(sr, sc): 0}
    min_len = INF
    while len(st) != 0:
        new_st = {}
        for (r, c), ln in st.items():
            for dr in [RT, DN, LT, UP]:
                r2, c2 = move(r, c, dr)
                if r2 >= 0 and c2 >= 0 and r2 < nr and c2 < nc:
                    ln2 = ln + d[r2][c2]
                    if (r2, c2) == ((nr-1), (nc-1)):
                        min_len = min(ln2, min_len)
                        #print("New min", min_len)
                        break
                    if ln2 < mn[r2][c2]:
                        new_st[(r2, c2)] = ln2
                        mn[r2][c2] = ln2
        st = new_st
    return min_len


def part1(d: List[List[int]]) -> int:
    return search(d)


def part2(d: List[List[int]]) -> int:
    nr, nc = len(d), len(d[0])
    NR, NC = 5*nr, 5*nc
    dd = [([0] * NC) for r in range(NR)]
    for i in range(5):
        for j in range(5):
            for r in range(nr):
                for c in range(nc):
                    v = (d[r][c] - 1 + i + j) % 9 + 1
                    dd[r + nr*i][c + nc*j] = v
    ln = search(dd)
    print(ln)
    return ln


assert part1(td) == 40
print("PART1:", part1(data))

assert part2(td) == 315
print("PART2:", part2(data))

print("Time for 3 runs:", timeit("part2(data)", number=3, globals=globals()))

File: 2021/15/202115.go
package main

import (
	"bytes"
	"fmt"
	"os"
)

func parse(s []byte) (g [][]int) {
	lines := bytes.Split(s, []byte("\n"))
	for _, line := range lines {
		r := make([]int, len(line))
		for c, b := range line {
			r[c] = int(b - 48)
		}
		g = append(g, r)
	}
	return
}

const INF = 999999999
const (
	RT = iota
	DN
	LT
	UP
)

type Point struct {
	r int
	c int
}

func move(p Point, dr int) Point {
	switch dr {
	case RT:
		return Point{p.r, p.c + 1}
	case DN:
		return Point{p.r + 1, p.c}
	case LT:
		return Point{p.r, p.c - 1}
	case UP:
		return Point{p.r - 1, p.c}
	}
	panic("wtf?")
}

func search(d [][]int) int {
	nr := len(d)
	nc := len(d[0])
	mn := [][]int{}
	for r := 0; r < nr; r++ {
		row := make([]int, nc)
		for c := 0; c < nc; c++ {
			row[c] = INF
		}
		mn = append(mn, row)
	}
	min_len := INF
	st := make(map[Point]int)
	st[Point{0, 0}] = 0
	for len(st) != 0 {
		new_st := make(map[Point]int)
		for p, ln := range st {
			for dr := 0; dr < 4; dr++ {
				p2 := move(p, dr)
				if p2.r >= 0 && p2.c >= 0 && p2.r < nr && p2.c < nc {
					ln2 := ln + d[p2.r][p2.c]
					if p2.r == nr-1 && p2.c == nc-1 {
						if ln2 < min_len {
							min_len = ln2
						}
						break
					}
					if ln2 < mn[p2.r][p2.c] {
						new_st[p2] = ln2
						mn[p2.r][p2.c] = ln2
					}
				}
			}
		}
		st = new_st
	}
	return min_len
}

func part1(d [][]int) int {
	return search(d)
}

func part2(d [][]int) int {
	nr1 := len(d)
	nc1 := len(d[0])
	nr2 := 5 * nr1
	nc2 := 5 * nc1

	dd := [][]int{}
	for r := 0; r < nr2; r++ {
		row := make([]int, nc2)
		dd = append(dd, row)
	}

	for i := 0; i < 5; i++ {
		for j := 0; j < 5; j++ {
			for r := 0; r < nr1; r++ {
				for c := 0; c < nc1; c++ {
					v := (d[r][c]-1+i+j)%9 + 1
					dd[r+nr1*i][c+nc1*j] = v
				}
			}
		}
	}
	ln := search(dd)
	return ln
}

var td []byte = []byte(
	`1163751742
1381373672
2136511328
3694931569
7463417111
1319128137
1359912421
3125421639
1293138521
2311944581`)

func main() {
	d := parse(td)
	t1 := part1(d)
	if t1 != 40 {
		panic("assertion failed")
	}

	dataStr, err := os.ReadFile("input.txt")
	if err != nil {
		panic("ReadFile")
	}
	data := parse(dataStr)

	ans1 := part1(data)
	fmt.Println("PART1:", ans1)

	t2 := part2(d)
	if t2 != 315 {
		panic("assertion t2 failed")
	}

	ans2 := part2(data)
	fmt.Println("PART2:", ans2)
}
